pgcontents/query.py

```diff
@@ -370,23 +371,74 @@ def rename_file(db, user_id, old_api_path, new_api_path):
             )
         )
 
-    # Overwriting existing files is disallowed.
-    if file_exists(db, user_id, new_api_path):
-        raise FileExists(new_api_path)
-
-    # Only allow writes to extant directories.
+    # Only allow writes to extant directories
     if not _dir_exists(db, user_id, new_dir):
         raise NoSuchDirectory(new_dir)
 
-    db.execute(
-        files.update().where(
-            _file_where(user_id, old_api_path),
-        ).values(
-            name=new_name,
-            created_at=func.now(),
+    # If we're renaming a file
+    if file_exists(db, user_id, old_api_path):
+        # Overwriting existing files is disallowed.
+        if file_exists(db, user_id, new_api_path):
+            raise FileExists(new_api_path)
+
+        db.execute(
+            files.update().where(
+                _file_where(user_id, old_api_path),
+            ).values(
+                name=new_name,
+                created_at=func.now(),
+            )
         )
-    )
 
+    # If we're renaming a directory
+    old_db_path = from_api_dirname(old_api_path)
+    new_db_path = from_api_dirname(new_api_path)
+    if _dir_exists(db, user_id, old_db_path):
```

Comment:
If the source directory doesn't exist, this should raise a NoSuchDirectory, not silently fail.